### ISSUE TITLE
Hetero R-GCN not matching OGB MAG Leaderboard performance

### ISSUE THREAD

**Issue by zjost** [dmlc/dgl#3511]
## 🐛 Bug
In an effort to directly reproduce the "NeighborSampling (R-GCN aggr)" results using DGL, I get test performance of `40.17 +/- 0.62`, compared to the PyG code linked on the leaderboard that gets `46.85 ± 0.48`

## To Reproduce

1. I adapted the example in the dgl examples directory: [`examples/pytorch/rgcn-hetero/entity_classify_mb.py`](https://github.com/dmlc/dgl/blob/master/examples/pytorch/rgcn-hetero/entity_classify_mb.py).  My implementation can be found [in this repo](https://github.com/zjost/rgcn-mag).  Run with `python rgcn_hetero_dgl.py`

## Expected behavior
I made an effort to exactly duplicate the PyG model architecture and parameter choices and I expect to get comparable performance.

## Environment
Training on single GPU.  

- DGL 0.7.0 (pip)
- PyTorch 1.9.0+cu111
- Python 3.6.9 
- Linux OS

**Comment by jermainewang** [comment]
Probably related to the effort in #3371

**Comment by zjost** [comment]
I have some additional information.  One thing the PyG implementation was doing differently is that instead of reversing `(paper, cites, paper)` into a new relation like `(paper, rev-cites, paper)`, it instead just made them into undirected edges ([code](https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppred/mag/sampler.py#L54)).  Making this change increased test accuracy from 40 to ~44.

Another significant difference is that in the dgl rgcn_hetero example, there is a single projection matrix for the `self.loop_weight` ([code](https://github.com/dmlc/dgl/blob/81915f55913a7a862f40fe2d8ea5b5319faf5bae/examples/pytorch/rgcn-hetero/model.py#L76)) that is shared for all self-connections, whereas in PyG there's a separate linear layer for each _node-type_ ([code](https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppred/mag/sampler.py#L106)).  To be explicit, there's a separate relational linear layer for each _edge-type_, and a separate "self" linear layer for each _node-type_.  With this setup, I'm able to match the same number of model parameters between PyG implementation and DGL.  However, with a trial of 10 runs, performance is as follows:
```
      train_acc  valid_acc  test_acc
mean   0.932635   0.457381  0.438895
std    0.001636   0.003455  0.003325
```

One peculiar thing I've noticed is that the _training_ accuracy in the case of PyG seems to reach a maximum of around 80%, but DGL is getting >90%.  This suggests a larger degree of over-fitting, but it doesn't appear that PyG has any other forms of regularization.

**Comment by zjost** [comment]
Another update: I'm now able to slightly beat PyG.  The only change was in capturing performance statistics in the same way.  Previously I was reporting performance at the end of the 3 epochs, but they actually report train/test performance of the epoch in which the validation accuracy was maximized ([code](https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppred/mag/logger.py#L17)).

However, there still remains a discrepancy:  the training accuracy is still significantly higher for DGL and it appears to overfit with fewer epochs.  This may be an artifact of differences in initialization methods.  I'll investigate this next.

```
Highest Train: 93.37 ± 0.20
Highest Valid: 49.03 ± 0.29
  Final Train: 76.15 ± 11.98
   Final Test: 47.66 ± 0.47
```

**Comment by zjost** [comment]
There were discrepancies between the methods used for initialization.  I have corrected this, but it did not have a significant impact on the results, particularly the training overfitting.

```
Highest Train: 93.68 ± 0.15
Highest Valid: 49.09 ± 0.48
  Final Train: 72.34 ± 11.93
   Final Test: 47.99 ± 0.68
```

Next I'll investigate the size of the gradients.

**Comment by zjost** [comment]
Interestingly, PyG seems to have larger gradients later in the training process than DGL.  Below I compare the average L2 norm of the gradient for each parameter Tensor across an epoch.  At the end of the first epoch, gradients are about the same magnitude:

![image](https://user-images.githubusercontent.com/9660566/142515141-e03f5fd5-4627-4b78-9ea7-916507945df5.png)

However, by the end of the 3rd epoch, gradients are significantly larger for PyG.

![image](https://user-images.githubusercontent.com/9660566/142515042-9611be8c-abb9-452b-9232-250bb00397f8.png)

This is likely just a reflection of the training loss differences.  For both DGL and PyG, the training loss average is about 2.3 at the end of epoch 1.  However, at the end of epoch 3, the training loss is 0.5 for DGL and 1.4 for PyG.  I assume this explains the larger gradients, but I'm still not clear why the training loss is so much lower for DGL.

**Comment by VoVAllen** [comment]
@zjost Hi, Could you try both setting with CPU? 

One possible reason can be PyG's cuda kernel uses atomic operation for reduction, which might be nondeterministic, and might hurt convergence. If testing on cpu, I think it can be less affected, at least not affected by the kernel implementation.

**Comment by mufeili** [comment]
By taking a look at your implementation and the PyG example, there can be a few additional potential implementation discrepancies：
- The `to_undirected` function [here](https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppred/mag/sampler.py#L54) will remove duplicate edges if any while your implementation does not.
- You passed `split_idx['train']` to `NodeDataLoader` while the PyG example passed `split_idx['train']['paper']` to `NeighborSampler`. It's likely `split_idx['train']` only contains the paper node IDs though.
- You did not call `model.train()` after the first time you called `test`.

**Comment by zjost** [comment]
Thanks for the tips!  I'll address results of each item:
* Using CPU vs GPU had no observable effect
* It's correct that I had duplicate edges (about 40k of them).  This did not have an observable effect
* `split_idx['train']` only includes `paper` nodes, so there is no difference that I can see
* It's correct that `model.train()` should have been called at the start of each epoch since the `test` function switched to `model.eval()` and effectively turned off dropout.  This reduced the training accuracy at the end of 3 epochs from 94 -> 85.
  * This also explains why things would be more or less the same after epoch 1, but then begin to overfit in subsequent epochs since dropout was removed.

The result of all this is that now the distribution of Training Accuracy at end of 3rd epoch is `84.38 ± 0.91` compared to PyG of `80.03 ± 0.40`.  This is still significantly significant, but much closer than before.

Next I'll try to manually run some forward/backward passes using a handful of nodes to see if there are any other clues.  Thanks again for all the great suggestions--I would have never found the "model.train" thing.

**Comment by zjost** [comment]
A few more discoveries.  If I manually set identical parameters between DGL and PyG models, and pass the same minibatch through (with all neighbors used, no sampling), I get identical outputs and gradients for all the parameters.  This suggests there's no fundamental computation problem in forward/backward pass.

One difference I've noticed is that the number of sampled neighbors do not match.  This seems to be because with a fanout of `[25, 20]`, DGL will sample that number of neighbors *per edge type*, whereas PyG samples that number and ignore edge type.  The consequence is that DGL touches many more nodes.  This likely results in more parameter updates in the training nodes and is likely resulting in faster over-fitting on the training nodes and convergence in general.

Quick question.  If I have a minibatch size of 1 and am using `fanout = [5, 2]`, I get something like this for the blocks of two layers:

```
>>> block[1]
Block(num_src_nodes={'author': 2, 'field_of_study': 2, 'institution': 0, 'paper': 3},
      num_dst_nodes={'author': 0, 'field_of_study': 0, 'institution': 0, 'paper': 1},
      num_edges={('author', 'affiliated_with', 'institution'): 0, ('author', 'writes', 'paper'): 2, ('field_of_study', 'rev-has_topic', 'paper'): 2, ('institution', 'rev-affiliated_with', 'author'): 0, ('paper', 'cites', 'paper'): 2, ('paper', 'has_topic', 'field_of_study'): 0, ('paper', 'rev-writes', 'author'): 0},
      metagraph=[('author', 'institution', 'affiliated_with'), ('author', 'paper', 'writes'), ('institution', 'author', 'rev-affiliated_with'), ('paper', 'paper', 'cites'), ('paper', 'field_of_study', 'has_topic'), ('paper', 'author', 'rev-writes'), ('field_of_study', 'paper', 'rev-has_topic')])

>>> block[0]
Block(num_src_nodes={'author': 25, 'field_of_study': 22, 'institution': 2, 'paper': 55},
      num_dst_nodes={'author': 4, 'field_of_study': 4, 'institution': 0, 'paper': 5},
      num_edges={('author', 'affiliated_with', 'institution'): 0, ('author', 'writes', 'paper'): 23, ('field_of_study', 'rev-has_topic', 'paper'): 25, ('institution', 'rev-affiliated_with', 'author'): 3, ('paper', 'cites', 'paper'): 21, ('paper', 'has_topic', 'field_of_study'): 20, ('paper', 'rev-writes', 'author'): 16},
      metagraph=[('author', 'institution', 'affiliated_with'), ('author', 'paper', 'writes'), ('institution', 'author', 'rev-affiliated_with'), ('paper', 'paper', 'cites'), ('paper', 'field_of_study', 'has_topic'), ('paper', 'author', 'rev-writes'), ('field_of_study', 'paper', 'rev-has_topic')])
```
I understand why `num_dst_nodes` for `block[1]` only contains the single node I'm generating a prediction for.  And the number of source nodes in that block contains, for example, 2 authors.  But my question is:  why is the `num_dst_nodes` equal to 4 authors in `block[0]`?  Shouldn't these be the same 2 authors from the downstream computation?

**Comment by zjost** [comment]
Final update.  I found that if I scaled the fan-out numbers in DGL by the number of (incoming) edge-types, I get about the same neighborhood sizes.  If I re-run 10 training iterations using `fanout = [6, 5]`, I get the following results:

```
All runs:
Highest Train: 80.09 ± 0.29
Highest Valid: 45.67 ± 0.44
  Final Train: 70.14 ± 6.29
   Final Test: 44.77 ± 0.55
```

This shows for the first time that the Training accuracy at the end of 3 epochs is matching PyG results, which adds credibility to the idea that the training data was overfitting more quickly in DGL because more nodes were involved in the computation and therefore getting embedding gradients.  

However, the Final Test accuracy has dropped significantly from 48% to 45%.  I suspect the difference is that sampling N nodes for each edge type is different than sampling `N*num_edge_types`, since in the first case edge-types will be represented equally whereas in the second the distribution of edge-types will reflect relative frequencies observed in the underlying graph.

Final question:  would there be value in me submitting a PR for this example that uses the Leaderboard architecture and achieves the higher final Test accuracy?
